biz.aQute.junit/src/aQute/junit/Activator.java

```diff
@@ -359,6 +371,46 @@ public void check() {
 		return -1;
 	}
 
+	/*
+	 * Find the host for a fragment. We just iterate over all other bundles and
+	 * ask for the fragments. We returns the first one found.
+	 */
+	private Bundle findHost(Bundle bundle) {
+		Bundle found = null;
+		PackageAdmin packageAdmin = (PackageAdmin) packageAdminTracker.getService();
+		if (packageAdmin == null) {
+			trace("Have a fragment but Package Admin not present to find the host");
+			return bundle;
+		}
+
+		for (Bundle potentialFragmentHost : context.getBundles()) {
+
+			if (potentialFragmentHost == bundle || potentialFragmentHost.getBundleId() == 0)
+				continue;
+
+			Bundle[] fragments = packageAdmin.getFragments(potentialFragmentHost);
+			if (fragments == null)
+				continue;
+
+			for (Bundle fragment : fragments) {
+				if (fragment == bundle) {
+					if (found != null) {
+						trace("Have a test fragment but find multiple hosts. Fragment=%s, Previous=%s, Next=%s ",
+								bundle, found, fragment);
+					} else
+						found = potentialFragmentHost;
+				}
+			}
+		}
+
+		return found == null ? bundle : found;
+	}
+
+	private boolean isFragment(Bundle bundle) {
+		return bundle != null && bundle.getState() != Bundle.ACTIVE
+				&& bundle.getHeaders().get(Constants.FRAGMENT_HOST) != null;
```

Comment:
Why don't you use `PackageAdmin.getBundleType`? You already track the package admin service now.